src/main/java/jenkins/branch/NameMangler.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2017, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.branch;
+
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Locale;
+import org.apache.commons.lang.StringUtils;
+
+/**
+ * Mangles names that are not nice so that they are safe to use on filesystem.
+ *
+ * @since 2.0.0
+ */
+public final class NameMangler {
+
+    private static final int MAX_SAFE_LENGTH = 32;
+    private static final int MIN_HASH_LENGTH = 6;
+    private static final int MAX_HASH_LENGTH = 12;
+
+    /**
+     * Utility class.
+     */
+    private NameMangler() {
+        throw new IllegalAccessError("Utility class");
+    }
+
+    public static String apply(String name) {
+        if (name.length() <= MAX_SAFE_LENGTH) {
+            boolean unsafe = false;
+            for (char c : name.toCharArray()) {
+                if (!isSafe(c)) {
+                    unsafe = true;
+                    break;
+                }
+            }
+            // See https://msdn.microsoft.com/en-us/library/aa365247 we need to consistently reserve names across all OS
+            if (!unsafe) {
```

Comment:
<iframe width="560" height="315" src="https://www.youtube.com/embed/dSIKBliboIo" frameborder="0" allowfullscreen></iframe>

I'm sorry Oleg I'm afraid I can't do that